### PR TITLE
Multinomial Logistic Regression datagen and Column-specific density support for sparse regression datagen

### PR DESCRIPTION
**Pull request by Er1cCheng**

### CODE REVIEW

**Comment by Er1cCheng** [comment]
build

File: python/benchmark/test_gen_data.py [hunk 1]
```diff
@@ -159,16 +163,43 @@ def test_make_regression(
...
+            assert c.shape == (col_num,), "coef shape mismatch"
+            assert np.count_nonzero(c) == 3, "Unexpected number of informative features"
+        else:
+            print(
+                c.shape,
+                (
+                    n_classes,
+                    col_num,
+                ),
+            )
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:184]: seems we can remove print statement. machine will not check it.

File: python/benchmark/test_gen_data.py [hunk 2]
```diff
@@ -159,16 +163,43 @@ def test_make_regression(
...
+            # Test that y consists of only discrete label
+            possible_labels = range(n_classes_num)
             for n in y:
-                assert n == 0 or n == 1
+                found = False
+                for l in possible_labels:
+                    if n == l:
+                        found = True
+                        break
+                assert found, "Invalid label"
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:202]: Have we already checked unique(y) == n_classes_num?

File: python/benchmark/test_gen_data.py [hunk 3]
```diff
@@ -177,17 +208,29 @@ def test_make_regression(
 @pytest.mark.parametrize("dtype", ["float64"])
 @pytest.mark.parametrize("use_gpu", ["True", "False"])
 @pytest.mark.parametrize("redundant_cols", ["0", "2"])
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:210]: thinking to mark some parameter values slow.

File: python/benchmark/test_gen_data.py [hunk 4]
```diff
@@ -258,7 +329,7 @@ def test_make_sparse_regression(
 
         # If there is no random shuffled redundant cols, we can check the total density
         density_num = float(density)
-        if redundant_cols == "0":
+        if redundant_cols == "0" and density_curve == "None":
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:332]: Seems only density_curve="None" is tested. How about density_curve="Linear" and "Exponential"?

File: python/benchmark/test_gen_data.py [hunk 5]
```diff
@@ -238,15 +286,38 @@ def test_make_sparse_regression(
...
+            assert c.shape == (col_num,), "coef shape mismatch"
+            assert np.count_nonzero(c) == 3, "Unexpected number of informative features"
+        else:
+            print(
+                c.shape,
+                (
+                    n_classes,
+                    col_num,
+                ),
+            )
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:300]: delete print statement

File: python/benchmark/test_gen_data.py [hunk 6]
```diff
@@ -238,15 +286,38 @@ def test_make_sparse_regression(
...
+            # Test that X consists of only discrete label
+            possible_labels = range(n_classes_num)
             for n in y:
-                assert n == 0 or n == 1
+                found = False
+                for l in possible_labels:
+                    if n == l:
+                        found = True
+                        break
+                assert found, "Invalid label"
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:320]: Can we check unique(y) == n_classes

File: python/benchmark/gen_data.py [hunk 7]
```diff
@@ -158,6 +158,8 @@ def _add_extra_parameters(self) -> None:
                 help_msg = "boolean for whether the shuffle the rows and cols of the feature matrix"
             elif name == "tail_strength":
                 help_msg = "tail strength for random low rank feature matrix generation, refer to sklearn.datasets.make_low_rank_matrix()"
+            elif name == "density_curve":
+                help_msg = "Specify columns wise density curve, support Linear or Exponential. Argument density would not be used to represent the max density in the curve"
```
Comment (lijinf2) [inline python/benchmark/gen_data.py:162]: What does Linear and Exponential mean?

Comment (lijinf2) [reply python/benchmark/gen_data.py:162]: what will be the output if density_curve is set to Linear or Exponential?

File: python/benchmark/gen_data_distributed.py [hunk 8]
```diff
@@ -695,26 +822,57 @@ def make_sparse_regression_udf(
...
-                    y = np.random.binomial(1, prob)
+                    if use_cupy:
+                        prob = 1 - 1 / (1 + cp.exp(-y_p))
+                        del y_p
+                        y = cp.random.binomial(1, prob)
+                    else:
+                        prob = 1 - 1 / (1 + np.exp(-y_p))
+                        del y_p
+                        y = np.random.binomial(1, prob)
+            else:
```
Comment (lijinf2) [inline python/benchmark/gen_data_distributed.py:874]: large chunk of code looks identical in dense class and sparse class. Can we merge identical code (e.g. create a python function)?

File: python/benchmark/test_gen_data.py (hunk 4 above)
Comment (Er1cCheng) [reply python/benchmark/test_gen_data.py:332]: For linear and exponential, the thing is that after shuffling the matrix seems just purely random. Is there a good way to test it?

File: python/benchmark/conftest.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    parser.addoption(
+        "--runslow", action="store_true", default=False, help="run slow tests"
+    )
+
+
+def pytest_configure(config: pytest.Config) -> None:
+    config.addinivalue_line("markers", "slow: mark as slow to run")
+
+
+def pytest_collection_modifyitems(
```
[the remark this answers is not in the document]
Comment (Er1cCheng) [reply python/benchmark/conftest.py:32]: The conftest file has to exist in the directory you want to control. This is a shrunked file compared to tests/conftest.py since we only need to support the slow option.

File: python/benchmark/test_gen_data.py (hunk 6 above)
Comment (Er1cCheng) [reply python/benchmark/test_gen_data.py:320]: There is the possibility that during randomness we lose completely a class even if less likely given sufficient data samples.

File: python/benchmark/gen_data_distributed.py (hunk 8 above)
Comment (Er1cCheng) [reply python/benchmark/gen_data_distributed.py:874]: Seems like an extra user defined function inside the UDF would break the pyspark dependency system. I get the same "Unable to find module gen_data_distribtued" error as last time.

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

File: python/benchmark/gen_data_distributed.py [hunk 10]
```diff
@@ -299,6 +299,41 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
...
+#         if use_cupy:
+#             prob = 1 - 1 / (1 + cp.exp(-y_p))
+#             del y_p
+#             y = cp.random.binomial(1, prob)
+#         else:
+#             prob = 1 - 1 / (1 + np.exp(-y_p))
+#             del y_p
+#             y = np.random.binomial(1, prob)
+
+#     return y
```
Comment (lijinf2) [inline python/benchmark/gen_data_distributed.py:334]: The comments should be deleted if not useful

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

File: python/benchmark/conftest.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (c) 2022, NVIDIA CORPORATION.
```
Comment (wbo4958) [inline python/benchmark/conftest.py:2]: 2022 -> 2024

File: python/run_benchmark.sh [hunk 12]
```diff
@@ -441,39 +441,42 @@ if [[ "${MODE}" =~ "logistic_regression" ]] || [[ "${MODE}" == "all" ]]; then
...
-            --num_cpus $num_cpus \
-            --num_runs $num_runs \
-            --train_path ${data_path} \
-            --transform_path ${data_path} \
-            --report_path "report_logistic_regression_${cluster_type}.csv" \
-            $common_confs $spark_rapids_confs \
-            ${EXTRA_ARGS}
+            if [[ ! -d ${data_path} ]]; then
+                python $gen_data_script sparse_regression \
+                --n_informative $( expr $num_cols / 3 )  \
```
Comment (eordentlich) [inline python/run_benchmark.sh:449]: should we have a new parameter num_sparse_cols instead of num_cols here?   We might want to make this much larger than the other dense cases.

File: python/benchmark/gen_data_distributed.py [hunk 13]
```diff
@@ -455,21 +487,63 @@ def make_regression_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
...
+                    if multinomial_log:
+                        if use_cupy:
+                            y = y.get()
+
+                        probs = [
+                            sp.special.softmax(target_weight) for target_weight in y
+                        ]
+
+                        multi_labels = [
+                            random.choices(range(n_classes), weights=p)[0]
```
[the remark this answers is not in the document]
Comment (Er1cCheng) [reply python/benchmark/gen_data_distributed.py:532]: Can't really find a way to achieve the matrix version, the softmax one does have a axis option

**Comment by Er1cCheng** [comment]
build

File: python/benchmark/gen_data.py [hunk 14]
```diff
@@ -164,6 +164,13 @@ def _add_extra_parameters(self) -> None:
             else:
                 help_msg = ""
 
+            # Support multiple biases
+            if name == "bias":
```
Comment (eordentlich) [inline python/benchmark/gen_data.py:168]: Why not generate bias randomly like the weights?

Comment (Er1cCheng) [reply python/benchmark/gen_data.py:168]: This is code inherited from before. Do we want random bias then?

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

**Comment by Er1cCheng** [comment]
build

File: python/benchmark/test_gen_data.py [hunk 15]
```diff
@@ -177,19 +219,49 @@ def test_make_regression(
...
+    [
+        ("None", "True"),
+        ("Linear", "True"),
+        ("Linear", "False"),
+        ("Exponential", "False"),
+        pytest.param("Exponential", "True", marks=pytest.mark.slow),
+    ],
+)
+@pytest.mark.parametrize(
+    "n_chunks", ["10", pytest.param("100", marks=pytest.mark.slow)]
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:249]: what is the relation of density, density_curve, and chunk?

**Comment by Er1cCheng** [comment]
build

File: python/benchmark/test_gen_data.py [hunk 16]
```diff
@@ -257,13 +365,70 @@ def test_make_sparse_regression(
...
+                )
+                density_values *= n_chunks_num * density_num / sum(density_values)
+
+            col_per_chunk = np.full(n_chunks_num, orig_cols // n_chunks_num)
+            col_per_chunk[: (orig_cols % n_chunks_num)] += 1
+            col_per_chunk = np.cumsum(col_per_chunk)
+
+            density_idx = 0
+            dense_count = 0
+            for idx, col in enumerate(X_np):
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:409]: what is the value of density_idx when the for loop is finished?

File: python/benchmark/test_gen_data.py [hunk 17]
```diff
@@ -257,13 +365,70 @@ def test_make_sparse_regression(
...
+                # Ignore the redundant columns
+                if idx >= orig_cols:
+                    break
+
+                if density_values[density_idx] == idx:
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:414]: not sure but it seems the if statement is always False? density_values[density_idx] is a float type, while idx is an int type?

**Comment by Er1cCheng** [comment]
build

File: python/benchmark/test_gen_data.py [hunk 18]
```diff
@@ -257,13 +365,70 @@ def test_make_sparse_regression(
...
+            num_partitions = 3
+
+            if isinstance(density, List):
+                density_num = float(density[0])
+            else:
+                density_num = float(density)
+
+            if density_curve == "Linear":
+                density_values = np.linspace(
+                    num_partitions / row_num, density_num, n_chunks_num
```
Comment (lijinf2) [inline python/benchmark/test_gen_data.py:392]: why density_values start from "num_partitions / row_num" not 0.? Can you give some explanation?

File: python/benchmark/test_gen_data.py (hunk 16 above)
Comment (lijinf2) [reply python/benchmark/test_gen_data.py:409]: please add the checking of the value after for loop is finished.

File: python/benchmark/test_gen_data.py (hunk 18 above)
Comment (Er1cCheng) [reply python/benchmark/test_gen_data.py:392]: The starting point for this is that I want to gaurantee the smallest density for each partition is big enough to generate at least one dense entry for the column. Thus, the density need to be at least 1/(rows/num_par) = num_par/rows. One potential issue is that num_partition is larger than row_num but that would defeat the whole purpose of this distributed datagen, so it should not happen at all. I wonder what's the "not 0." issue?

**Comment by Er1cCheng** [comment]
build

**Review by lijinf2 — APPROVED** [review]
Looks nice!
